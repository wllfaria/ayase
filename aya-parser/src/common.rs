use nom::branch::alt;
use nom::bytes::complete::{tag, tag_no_case};
use nom::character::complete::{alpha1, alphanumeric1, anychar, char, hex_digit1, multispace0, one_of};
use nom::combinator::{map, peek as nom_peek, recognize};
use nom::multi::many0_count;
use nom::sequence::{pair, terminated};
use nom::IResult;

use crate::types::{Atom, Operator};

pub mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 1;
    pub const MUL: u8 = 2;
}

pub fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Add | Operator::Sub => precedences::SUM,
        Operator::Mul => precedences::MUL,
    }
}

pub fn is_terminator(ch: char) -> bool {
    matches!(ch, ']' | ')')
}

pub fn peek(input: &str) -> IResult<&str, char> {
    nom_peek(anychar)(input)
}

pub fn general_register(input: &str) -> IResult<&str, &str> {
    recognize(pair(tag_no_case("r"), one_of("12345678")))(input)
}

pub fn register(input: &str) -> IResult<&str, Atom> {
    map(
        alt((
            general_register,
            tag_no_case("ip"),
            tag_no_case("sp"),
            tag_no_case("fp"),
            tag_no_case("ret"),
        )),
        Atom::Register,
    )(input)
}

pub fn identifier(input: &str) -> IResult<&str, &str> {
    recognize(pair(
        alt((alpha1, tag("_"))),
        many0_count(alt((alphanumeric1, tag("_")))),
    ))(input)
}

pub fn variable(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("!")(input)?;
    map(identifier, Atom::Var)(input)
}

pub fn label(input: &str) -> IResult<&str, Atom> {
    let (input, _) = multispace0(input)?;
    let (input, label) = map(terminated(identifier, char(':')), Atom::Label)(input)?;
    let (input, _) = multispace0(input)?;
    Ok((input, label))
}

pub fn hex_literal(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("$")(input)?;
    map(hex_digit1, Atom::HexLiteral)(input)
}

pub fn address(input: &str) -> IResult<&str, Atom> {
    let (input, _) = tag("&")(input)?;
    map(hex_digit1, Atom::Address)(input)
}

pub fn operator(input: &str) -> IResult<&str, Atom> {
    map(
        alt((
            map(char('+'), |_| Operator::Add),
            map(char('-'), |_| Operator::Sub),
            map(char('*'), |_| Operator::Mul),
        )),
        Atom::Operator,
    )(input)
}
