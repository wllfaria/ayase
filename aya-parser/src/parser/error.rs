use crate::ast::ByteOffset;
use crate::lexer::Token;

pub type Result<T> = std::result::Result<T, miette::Error>;

pub static ADDRESS_HELP: &str = "valid addresses takes the form of &FFFF";
pub static ADDRESS_MSG: &str = "[SYNTAX_ERROR]: expected address";

pub static HEX_LIT_HELP: &str = "valid hex literals takes the form of $FFFF";
pub static HEX_LIT_MSG: &str = "[SYNTAX_ERROR]: expected hex literal";

pub static PATH_MSG: &str = "[SYNTAX_ERROR]: expected path string";

pub static IDENT_MSG: &str = "[SYNTAX_ERROR]: expected valid identifier";

pub static EOF_MSG: &str = "[SYNTAX_ERROR]: unexpected end of file [EOF]";

pub static UNEXPECTED_TOKEN_MSG: &str = "[SYNTAX_ERROR]: unexpected token";

pub static COLON_MSG: &str = "[SYNTAX_ERROR]: expected a `:` [COLON]";
pub static DOT_MSG: &str = "[SYNTAX_ERROR]: expected a `.` [DOT]";
pub static PLUS_MSG: &str = "[SYNTAX_ERROR]: expected a `+` [PLUS]";
pub static COMMA_MSG: &str = "[SYNTAX_ERROR]: expected a `,` [COMMA]";
pub static LBRACE_MSG: &str = "[SYNTAX_ERROR]: expected a `{` [LEFT_CURLY]";
pub static RBRACE_MSG: &str = "[SYNTAX_ERROR]: expected a `}` [RIGHT_CURLY]";
pub static LBRACKET_MSG: &str = "[SYNTAX_ERROR]: expected a `[` [LEFT_BRACKET]";
pub static RBRACKET_MSG: &str = "[SYNTAX_ERROR]: expected a `]` [RIGHT_BRACKET]";

pub fn bail<S: AsRef<str>>(source: S, help: S, message: S, offset: impl Into<ByteOffset>) -> miette::Error {
    let offset = offset.into();
    miette::Error::from(
        miette::MietteDiagnostic::new(message.as_ref())
            .with_labels(vec![miette::LabeledSpan::at(offset.start..offset.end, "this bit")])
            .with_help(help.as_ref()),
    )
    .with_source_code(source.as_ref().to_string())
}

pub fn unexpected_eof<S: AsRef<str>, T>(source: S, help: S) -> Result<T> {
    let end = source.as_ref().len();
    let start = end.saturating_sub(1);
    Err(bail(source.as_ref(), help.as_ref(), EOF_MSG, start..end))
}

pub fn unexpected_token<S: AsRef<str>, T>(source: S, token: &Token) -> Result<T> {
    Err(bail(
        source.as_ref(),
        &format!("unexpected token {token}"),
        UNEXPECTED_TOKEN_MSG,
        token.offset(),
    ))
}
