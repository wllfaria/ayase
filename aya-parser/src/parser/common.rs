use std::ops::Range;

use super::error::{bail, unexpected_eof, unexpected_token};
use super::Result;
use crate::ast::ByteOffset;
use crate::lexer::{Kind, Lexer, Token, TransposeRef};

pub fn expect<S: AsRef<str>>(
    expected: Kind,
    lexer: &mut Lexer,
    source: S,
    help: S,
    message: S,
) -> miette::Result<ByteOffset> {
    let current = lexer.next().transpose()?;
    let Some(current) = current else {
        unreachable!();
    };

    if current.kind != expected {
        return Err(bail(source.as_ref(), help.as_ref(), message.as_ref(), current.offset()));
    }

    Ok(current.offset())
}

pub fn expect_fail<S: AsRef<str>>(expected: Kind, lexer: &mut Lexer, source: S) -> miette::Result<bool> {
    let Ok(Some(next)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    if next.kind != expected {
        return unexpected_token(source.as_ref(), next);
    }

    lexer.next().transpose()?;
    Ok(false)
}

pub fn parse_identifier<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::Ident, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_register<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<ByteOffset> {
    let offset = parse_identifier(source.as_ref(), lexer, "", "")?;
    let name = &source.as_ref()[Range::<usize>::from(offset)];
    match name {
        "acc" | "ip" | "r1" | "r2" | "r3" | "r4" | "r5" | "r6" | "r7" | "r8" | "sp" | "fp" => Ok(offset),
        _ => unexpected_token(source.as_ref(), &Token::from_ident(name, offset.start, offset.end)),
    }
}

pub fn parse_hex_lit<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::HexNumber, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_string<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::String, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_variable<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::Bang, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    expect(Kind::Ident, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_keyword<S: AsRef<str>>(source: S, lexer: &mut Lexer, expected: Kind) -> Result<ByteOffset> {
    let next = lexer.next().transpose()?;
    let Some(next) = next else {
        return unexpected_eof(source.as_ref(), "invalid eof");
    };

    if next.kind != expected {
        return unexpected_token(source.as_ref(), &next);
    }

    Ok(next.offset())
}
