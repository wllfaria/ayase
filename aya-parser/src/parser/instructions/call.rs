use crate::ast::{Instruction, Statement};
use crate::lexer::{Kind, Lexer, TransposeRef};
use crate::parser::common::{parse_hex_lit, parse_keyword, parse_variable};
use crate::parser::error::{unexpected_eof, unexpected_token, HEX_LIT_HELP, HEX_LIT_MSG, IDENT_MSG};
use crate::parser::syntax::parse_simple_address;
use crate::parser::Result;

pub fn parse_call<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    parse_keyword(source.as_ref(), lexer, Kind::Call)?;

    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };
    let kind = token.kind;

    let value = match kind {
        Kind::HexNumber => Statement::Address(parse_hex_lit(source.as_ref(), lexer, HEX_LIT_HELP, HEX_LIT_MSG)?),
        Kind::Ampersand => parse_simple_address(source.as_ref(), lexer, HEX_LIT_HELP, HEX_LIT_MSG)?,
        Kind::Bang => Statement::Var(parse_variable(
            source.as_ref(),
            lexer,
            "variable must be a valid identifier",
            IDENT_MSG,
        )?),
        _ => return unexpected_token(source.as_ref(), token),
    };

    match kind {
        Kind::HexNumber => Ok(Instruction::CallLit(value).into()),
        Kind::Ampersand => Ok(Instruction::CallLit(value).into()),
        Kind::Bang => Ok(Instruction::CallLit(value).into()),
        _ => unreachable!(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> Statement {
        let mut lexer = Lexer::new(input);
        parse_call(input, &mut lexer).unwrap()
    }

    #[test]
    fn test_call_address() {
        let input = "call &[$0303]";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_call_lit() {
        let input = "call $0303";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }
}
