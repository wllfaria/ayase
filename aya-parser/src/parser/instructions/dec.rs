use crate::ast::{Instruction, Statement};
use crate::lexer::{Kind, Lexer};
use crate::parser::common::{parse_keyword, parse_register};
use crate::parser::Result;

pub fn parse_dec<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    parse_keyword(source.as_ref(), lexer, Kind::Dec)?;
    let value = Statement::Register(parse_register(source.as_ref(), lexer)?);
    Ok(Instruction::Dec(value).into())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> Statement {
        let mut lexer = Lexer::new(input);
        parse_dec(input, &mut lexer).unwrap()
    }

    #[test]
    fn test_dec_reg() {
        let input = "dec r1";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }
}
