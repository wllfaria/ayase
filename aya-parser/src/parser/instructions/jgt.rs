use crate::ast::{Instruction, Statement};
use crate::lexer::{Kind, Lexer, TransposeRef};
use crate::parser::common::{expect, parse_hex_lit, parse_keyword, parse_register};
use crate::parser::error::{
    unexpected_eof, unexpected_token, ADDRESS_HELP, ADDRESS_MSG, COMMA_MSG, HEX_LIT_HELP, HEX_LIT_MSG,
};
use crate::parser::syntax::parse_simple_address;
use crate::parser::Result;

pub fn parse_jgt<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    parse_keyword(source.as_ref(), lexer, Kind::Jgt)?;

    let lhs = parse_simple_address(source.as_ref(), lexer, ADDRESS_HELP, ADDRESS_MSG)?;

    expect(
        Kind::Comma,
        lexer,
        source.as_ref(),
        "missing a comma after left side of instruction",
        COMMA_MSG,
    )?;

    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    let kind = token.kind;
    let rhs = match kind {
        Kind::Ident => Statement::Register(parse_register(source.as_ref(), lexer)?),
        Kind::HexNumber => Statement::HexLiteral(parse_hex_lit(source.as_ref(), lexer, HEX_LIT_HELP, HEX_LIT_MSG)?),
        _ => return unexpected_token(source.as_ref(), token),
    };

    match kind {
        Kind::Ident => Ok(Instruction::JgtReg(lhs, rhs).into()),
        Kind::HexNumber => Ok(Instruction::JgtLit(lhs, rhs).into()),
        _ => unreachable!(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> Statement {
        let mut lexer = Lexer::new(input);
        parse_jgt(input, &mut lexer).unwrap()
    }

    #[test]
    fn test_jgt_reg() {
        let input = "jgt &[$c0d3], r2";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_jgt_lit() {
        let input = "jgt &[$c0d3], $0303";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }
}
