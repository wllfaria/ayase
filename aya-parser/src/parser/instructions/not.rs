use crate::ast::{Instruction, Statement};
use crate::lexer::{Kind, Lexer};
use crate::parser::common::{parse_keyword, parse_register};
use crate::parser::Result;

pub fn parse_not<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    parse_keyword(source.as_ref(), lexer, Kind::Not)?;
    let value = Statement::Register(parse_register(source.as_ref(), lexer)?);
    Ok(Instruction::Not(value).into())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> Statement {
        let mut lexer = Lexer::new(input);
        parse_not(input, &mut lexer).unwrap()
    }

    #[test]
    fn test_not_reg() {
        let input = "not r1";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }
}
