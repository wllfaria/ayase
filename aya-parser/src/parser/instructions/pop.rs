use crate::ast::{Instruction, Statement};
use crate::lexer::{Kind, Lexer};
use crate::parser::common::{parse_keyword, parse_register};
use crate::parser::Result;

pub fn parse_pop<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    parse_keyword(source.as_ref(), lexer, Kind::Pop)?;
    let value = Statement::Register(parse_register(source.as_ref(), lexer)?);
    Ok(Instruction::Pop(value).into())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> Statement {
        let mut lexer = Lexer::new(input);
        parse_pop(input, &mut lexer).unwrap()
    }

    #[test]
    fn test_pop_reg() {
        let input = "pop r2";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }
}
