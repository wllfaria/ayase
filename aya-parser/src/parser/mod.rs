mod common;
pub mod error;
mod import;
mod instructions;
mod syntax;

use common::expect;
pub use error::Result;
use error::{unexpected_eof, unexpected_token, PLUS_MSG};
use import::*;
use instructions::*;
use syntax::*;

use crate::ast::{Ast, Statement};
use crate::lexer::{Kind, Lexer, TransposeRef};

fn parse_instruction<S: AsRef<str>>(source: S, lexer: &mut Lexer, kind: Kind) -> Result<Statement> {
    match kind {
        Kind::Mov => parse_mov(source, lexer),
        Kind::Add => parse_add(source, lexer),
        Kind::Sub => parse_sub(source, lexer),
        Kind::Mul => parse_mul(source, lexer),
        Kind::Lsh => parse_lsh(source, lexer),
        Kind::Rsh => parse_rsh(source, lexer),
        Kind::And => parse_and(source, lexer),
        Kind::Or => parse_or(source, lexer),
        Kind::Xor => parse_xor(source, lexer),
        Kind::Inc => parse_inc(source, lexer),
        Kind::Dec => parse_dec(source, lexer),
        Kind::Not => parse_not(source, lexer),
        Kind::Jeq => parse_jeq(source, lexer),
        Kind::Jgt => parse_jgt(source, lexer),
        Kind::Jne => parse_jne(source, lexer),
        Kind::Jge => parse_jge(source, lexer),
        Kind::Jle => parse_jle(source, lexer),
        Kind::Jlt => parse_jlt(source, lexer),
        Kind::Psh => parse_psh(source, lexer),
        Kind::Pop => parse_pop(source, lexer),
        Kind::Call => parse_call(source, lexer),
        Kind::Ret => parse_ret(source, lexer),
        Kind::Hlt => parse_hlt(source, lexer),
        _ => unreachable!(),
    }
}

fn parse_exported_identifier<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    expect(Kind::Plus, lexer, source.as_ref(), "expected a `+` [PLUS]", PLUS_MSG)?;

    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    match token.kind {
        Kind::Ident => parse_label(source, lexer, true),
        Kind::Data8 => parse_data(source.as_ref(), lexer, DataSize::Byte, true),
        Kind::Data16 => parse_data(source.as_ref(), lexer, DataSize::Word, true),
        Kind::Const => parse_const(source.as_ref(), lexer, true),
        _ => unexpected_token(source.as_ref(), token),
    }
}

fn parse_statement<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };
    let kind = token.kind;
    match kind {
        Kind::Import => parse_import(source, lexer),
        Kind::Plus => parse_exported_identifier(source, lexer),
        Kind::Data8 => parse_data(source.as_ref(), lexer, DataSize::Byte, false),
        Kind::Data16 => parse_data(source.as_ref(), lexer, DataSize::Word, false),
        Kind::Const => parse_const(source, lexer, false),
        Kind::Ident => parse_label(source, lexer, false),
        k if k.is_instruction() => parse_instruction(source, lexer, kind),
        _ => unexpected_token(source.as_ref(), token),
    }
}

pub fn parse<S: AsRef<str>>(source: S) -> Result<Ast> {
    set_miette_hook();

    let source = source.as_ref();
    let mut lexer = Lexer::new(source);
    let mut statements = vec![];

    while !lexer.is_empty() {
        let statement = parse_statement(source, &mut lexer)?;
        statements.push(statement);
    }

    Ok(Ast { statements })
}

fn set_miette_hook() {
    miette::set_hook(Box::new(|_| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .terminal_links(true)
                .context_lines(10)
                .tab_width(2)
                .color(true)
                .break_words(true)
                .build(),
        )
    }))
    .ok();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_private_label() {
        let input = "name:";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_exported_label() {
        let input = "+name:";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_private_constant() {
        let input = "const NAME = $0123";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_exported_constant() {
        let input = "+const NAME = &[$0123]";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_private_data8() {
        let input = "data8 NAME = { &[$0123], $1234 }";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_exported_data8() {
        let input = "+data8 NAME = { &[$0123], $1234, }";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_private_data16() {
        let input = "data16 NAME = { &[$0123], $1234 }";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_exported_data16() {
        let input = "+data16 NAME = { &[$0123], $1234, }";
        let result = parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }
}
