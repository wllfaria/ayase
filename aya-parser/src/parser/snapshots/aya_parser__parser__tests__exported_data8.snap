---
source: aya-parser/src/parser/mod.rs
expression: result
---
Ast {
    statements: [
        Data {
            name: ByteOffset {
                start: 7,
                end: 11,
            },
            size: 8,
            exported: true,
            values: [
                Address(
                    ByteOffset {
                        start: 19,
                        end: 23,
                    },
                ),
                HexLiteral(
                    ByteOffset {
                        start: 27,
                        end: 31,
                    },
                ),
            ],
        },
    ],
}
